sound/soc/intel/boards/sof_sdw.c

```diff
@@ -1011,6 +1011,8 @@ static int get_slave_info(const struct snd_soc_acpi_link_adr *adr_link,
 	return 0;
 }
 
+static const char * const type_strings[] = {"SimpleJack", "SmartAmp", "SmartMic"};
+
```

Comment:
> @bardliao I am not following the comment "That won't impact the existing topology with this way."
> 
> Are you saying that the existing topology can be used even for this case? I don't see how because the dailinks are different?
> 
> Or that we need to generate a new topology where the second dailink name is modified? If yes, can you share the topology change?

@plbossart I was trying to say the existing topologies still work with the same HW configurations. Because this commit doesn't change the existing dai link name.
For example, the dai link names are still
```
SDW0-Playback
SDW0-Capture
SDW1-Playback
SDW1-Capture
SDW3-Capture
```
They won't become
```
SDW0-Playback-SimpleJack
SDW0-Capture-SimpleJack
SDW1-Playback-SmartAmp
SDW1-Capture-SmartAmp
SDW3-Capture-SmartMic
```
We only append the type if a previous codec is on the same bus.
For example, If there are two codecs on SDW 0 bus, the dai names will be
```
SDW0-Playback
SDW0-Capture
SDW0-Playback-SmartAmp
SDW0-Capture-SmartAmp
SDW3-Capture
```
The link name will only be changed in such new HW configurations. That's why I said "That won't impact the existing topologies"

Reply:
You will still need to modify the topology generation to sometimes use SDW0-Playback-SmartAmp and most times use SDW1-Playback, and in addition the list of dailinks looks weird. How do we know is SDW0-Capture is a jack or a mic?

It's almost like we need to move to a different convention with more explicit names for all links, with some sort of backwards compatibility alias, instead of adding the alias in some cases only.

Need to think more on this one.

Reply:
> You will still need to modify the topology generation to sometimes use SDW0-Playback-SmartAmp and most times use SDW1-Playback, and in addition the list of dailinks looks weird. How do we know is SDW0-Capture is a jack or a mic?

We only need to add a new topology uses SDW0-Playback-SmartAmp when a product ues jack and amp on the same sdw bus, right? 
For example,
```
"cavs-sdw\;sof-tgl-rt712\;SDW_SPK_STREAM=SDW0-Playback-SmartAmp"
```
The link order is from `struct snd_soc_acpi_link_adr` and we have suggested the jack -> amp -> mic order. Kernel will show a warning if the order doesn't follow the suggestion. So, `SDW0-Capture` before `SDW0-Capture-SmartAmp` should be a jack link.

> 
> It's almost like we need to move to a different convention with more explicit names for all links, with some sort of backwards compatibility alias, instead of adding the alias in some cases only.

That sounds great. But I am not sure do we have backwards compatibility alias. Do you have any idea how to have backwards compatibility alias?

> 
> Need to think more on this one.

Reply:
We probably need to ask @ranj063 to remind us how the dailink is bound between machine driver and topology, is this something that we could update with multiple levels and fallbacks?

Reply:
> We probably need to ask @ranj063 to remind us how the dailink is bound between machine driver and topology, is this something that we could update with multiple levels and fallbacks?

It may work if we can check is there are different codecs on the same sdw bus first, then we can decide what kind of link name format we need.

Reply:
> > We probably need to ask @ranj063 to remind us how the dailink is bound between machine driver and topology, is this something that we could update with multiple levels and fallbacks?
> 
> It may work if we can check is there are different codecs on the same sdw bus first, then we can decide what kind of link name format we need.

@plbossart Can you review the new version? I implemented my thought. If there are two different codecs on the same sdw link, the driver will append codec type to ALL sdw dai link names.